### ISSUE TITLE
Generics issues

### ISSUE THREAD

**Issue by phadej** [gcanti/flowcheck#10]
There are few (probably unrelated) issues, but they are triggered by same example:

```js
class Person<A> {
  constructor(name: A) {
    this.name = name;
  }
}

var p: Person = new Person<String>(41);
var n: number = p.name;
console.log(n);
```

generates

```js
function Person(name) {
  "use strict";
  f.check(arguments, f.arguments([f.any]));
  this.name = name;
}


var p = f.check(new Person < String > (41), Person); // 1. and 2.
var n = f.check(p.name, f.number); 
console.log(n);
```

1. the `< Person >` isn't stripped, even there is `stripTypes` enabled - and looks like valid JS!
2. flowcheck doesn't check that `String` type is present. Actually with correct lowercased `string`, the JS won't work because of **1.**

I'd expect the generated code to be something like:

```js
var p = f.check(new Person(41), Person.type.instantiate(f.string)); // will fail as 41 is not string.
```

**Comment by phadej** [comment]
Given a bit longer though, I don't even know how the correct implementation could look like. The `new Person(41)` could look like `new Person.instantiated(f.string)(41)`, which will then fail. But I have no idea how to make `instanceof` and arguments checks work for it together.

Or you have to admit (and document) that polymorphic types are erased at the runtime (as e.g. in Java).

**Comment by gcanti** [comment]
Hi, thanks for your feedback.
After a quick test I think there is a problem with polymorphic classes and stripTypes (flowcheck disabled):

input

```js
var p: Person = new Person<string>(41);
```

output of JSXTransformer:

```js
var p         = new Person<string>(41);
```

output of 6to5:

```js
"use strict";

var p = new Person() < string > 41;
```

> Or you have to admit (and document) that polymorphic types are erased at the runtime (as e.g. in Java).

Absolutely, I'll add a warning in README.md, thanks.

**Comment by phadej** [comment]
Ah seems that `new Person<string>(41)` doesn't do what I'd expect even in flow. And if I try:
```js
/* @flow */

class Person<A> {
  constructor(name: A) {
    this.name = name;
  }
}

type T = Person<String>;
var p: T = new Person(41);
```

flow (0.1.2) doesn't complain either, strange.

**Comment by phadej** [comment]
Feel free to close after you add a warning in README.md.
Looks like that
```js
/* @flow */

class Person<A> {
  constructor(name: A) {
    this.name = name;
  }
}

type T = Person<string>;
var p: T = new Person(41);
```
produces what I'd expect, if types are erased.

**Comment by gcanti** [comment]
flow 0.1.4 as well

Besides, JSXTransformer with stripTypes enabled translates

```js
class Person<A> {
  constructor(name: A) {
    this.name = name;
  }
}
```

to

```js
class Person < A > {
  constructor(name) {
    this.name = name;
  }
}
```

Definitely there are issues with stripTypes and generics

**Comment by phadej** [comment]
Looks like that `6to5` strips `< A >`:
```
function Person(name) {
  "use strict";
  f.check(arguments, f.arguments([f.any]));
  this.name = name;
}
```

:)
